test/environment_unittest.cc

```diff
@@ -0,0 +1,102 @@
+#include "../src/environment.h"
+#include "gtest/gtest.h"
+
+#include <fstream>
+#include <sstream>
+#include <boost/filesystem.hpp>
+
+namespace google {
+
+class EnvironmentTest : public ::testing::Test {
+ protected:
+  static void ReadApplicationDefaultCredentials(const Environment& environment) {
+    environment.ReadApplicationDefaultCredentials();
+  }
+};
+
+namespace {
+// A file with a given name in a temporary (unique) directory.
+boost::filesystem::path TempPath(const std::string& filename) {
+  boost::filesystem::path path = boost::filesystem::temp_directory_path();
+  path.append(boost::filesystem::unique_path().native());
+  path.append(filename);
+  return path;
+}
+
+// Creates a file for the lifetime of the object and removes it after.
+class TemporaryFile {
+ public:
+  TemporaryFile(const std::string& filename, const std::string& contents)
+      : path_(TempPath(filename)) {
+    boost::filesystem::create_directories(path_.parent_path());
+    SetContents(contents);
+  }
+  ~TemporaryFile() {
+    boost::filesystem::remove_all(path_.parent_path());
+  }
+  void SetContents(const std::string& contents) const {
+    std::ofstream file(path_.native());
+    file << contents << std::flush;
+  }
+  const std::string& FullPath() const { return path_.native(); }
+ private:
+  boost::filesystem::path path_;
+};
+}  // namespace
+
+TEST(TemporaryFile, Basic) {
+  std::string path;
+  {
+    TemporaryFile f("foo", "bar");
+    path = f.FullPath();
+    EXPECT_TRUE(boost::filesystem::exists(path));
+    std::string contents;
+    {
+      std::ifstream in(path);
+      in >> contents;
+    }
+    EXPECT_EQ("bar", contents);
+    f.SetContents("xyz");
+    {
+      std::ifstream in(path);
+      in >> contents;
+    }
+    EXPECT_EQ("xyz", contents);
+  }
+  EXPECT_FALSE(boost::filesystem::exists(path));
+}
+
+TEST_F(EnvironmentTest, ReadApplicationDefaultCredentialsSucceeds) {
+  TemporaryFile credentials_file(
+    std::string(test_info_->name()) + "_creds.json",
```

Comment:
where does test_info_ come from?

Reply:
test_info_ is provided by gtest.

Reply:
Umm, yeah, except the [preferred way to access it](https://github.com/google/googletest/blob/master/googletest/docs/AdvancedGuide.md#getting-the-current-tests-name) seems to be by calling `current_test_info()` on `UnitTest`. I'll send a separate PR with a fix.